solve_univariate_inequality: solving inequality with imaginary coefficients

akash9712 opened the pull request:
Added support for solving inequalities involving rational functions with
imaginary coefficients. solve_univariate_inequality() has been modified to
return those values of the generator for which the function assumes a real
value and satisfies the inequality. Tests have been added.

Examples:
```
>>> isolve = solve_univariate_inequality
>>> eq = (x + I)*(x + 2*I) < 0
>>> isolve(eq, x)
Eq(x, 0)
>>> x12 = (x - 1)*(x - 2)
>>> isolve(eq.subs(x, x12), x)
Eq(x, 1) | Eq(x, 2)
```

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->
Fixes #13105

akash9712 committed:
solve_univariate_inequalities: solving inequality with imaginary coefficients

Added support for solving inequalities involving rational functions with
imaginary coefficients. solve_univariate_inequality has been modified to
return those values of the generator for which the function assumes a real
value and satisfies the inequality. Tests have been added.

Examples:

>>> isolve = solve_univariate_inequality
>>> eq = (x + I)*(x + 2*I) < 0
>>> isolve(eq, x)
Eq(x, 0)
>>> x12 = (x - 1)*(x - 2)
>>> isolve(eq.subs(x, x12), x)
Eq(x, 1) | Eq(x, 2)

akash9712 committed:
solve_univariate_inequalities: solving inequality with imaginary coefficients

Added support for solving inequalities involving rational functions with
imaginary coefficients. solve_univariate_inequalities has been modified to
return those values of the generator for which the function assumes a real
value and satisfies the inequality. Tests have been added.

Examples:

>>> isolve = solve_univariate_inequality
>>> eq = (x + I)*(x + 2*I) < 0
>>> isolve(eq, x)
Eq(x, 0)
>>> x12 = (x - 1)*(x - 2)
>>> isolve(eq.subs(x, x12), x)
Eq(x, 1) | Eq(x, 2)

ArifAhmed1995 commented on sympy/solvers/inequalities.py:
```diff
@@ -603,7 +623,11 @@ def valid(x):
-            rv = Union(*sol_sets).subs(gen, _gen)
+            if expanded_e.coeff(I) != S.Zero:
+                rv = (make_real)
+
```
@akash9712 This extra blank line is not needed.

akash9712 replied:
@ArifAhmed1995 I added that line to seperate the if clauses, as done before this pull request. Are you sure it should be removed?

ArifAhmed1995 replied:
Yeah, I think blank line `628` should be removed.

akash9712 committed:
Removed blank lines

Removed a blank line after an if clause.

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into 13105_inequality

akash9712 committed:
Removed blank lines

Removed a blank line after an if clause.

smichr commented:
This maybe should only be done if `expr.rel_op not in '!=='`; in case of Eq and Ne, imaginary solutions are permitted. What does this PR give for `isolve(Eq((x+I)*(x+2*I), -6), x)`? It should give `[-4*I, I]`.

akash9712 commented:
The PR would return `False` for isolve `(Eq((x+I)*(x+2*I), -6), x)`.
https://github.com/sympy/sympy/blob/master/sympy/solvers/inequalities.py#L449
This line suggests that the function should be called only for real domains. Should the function be changed to return complex values for the generator?

smichr commented:
> Should the function be changed to return complex values for the generator?

No. Thanks for the reminder. solveset already solves the equation correctly. There is no need to call this function when dealing with an Eq instance.

smichr commented:
Did you consider splitting the equation into real and imaginary parts rather than explicitly using "I"? What happens if you replace `I` with `Symbol("I", imaginary=True)`?

akash9712 commented:
I tried using the condition `im(expr) == 0` to check if the function contains an imaginary part, but the condition returns `False` in some cases even when the function would be real on the domain given by `continuous_domain`. For eg. 
```
>>> continuous_domain(log(x), x, S.Reals)
Interval.open(0, oo)
>>> im(log(x)) == 0
False
>>> im(log(x))
arg(x)
```

akash9712 commented:
@smichr I have tried working around this issue, but I needed to make certain changes to the `solveset` function for the solving of any function of type `arg`. Should I open a seperate pull request for that?

akash9712 committed:
solve_univariate_inequality: Changing how imaginary part is handled

Changes were made to `solve_univariate_inequality` function. Function now
accepts arguments with non zero imaginary coefficients as input and
solves them for zero, returning real values satisfying the inequality if any.
Also added support for solving type `arg` in `solveset`.

akash9712 committed:
solve_univariate_inequality: Modified function and added tests

Changes were made to consider the boundary conditions for solving
inequalities which evaluating to Intervals. Added tests to for solveset
regarding changes made to `solveset` for evaluating equalities of type `arg`.

Examples for modification to solveset:

>>> solveset(arg(x), x)
Interval.open(0, oo)
>>> solveset(arg(4*x-3), x)
Interval.open(3/4, oo)

akash9712 commented:
@smichr Please review the changes.

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -571,6 +573,47 @@ def valid(x):
+                                im_sol += FiniteSet(z)
+                    else:
+                        start, end = a.inf, a.sup
+                        for z in _nsort(critical_points + FiniteSet(end)):
```
Recommend that you `compute valid_start = valid(start)` once here after defining start and then use that variable below.

smichr replied:
Recommend that you compute `valid_z = valid(z)` immeadiately after starting the z loop.

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -571,6 +573,47 @@ def valid(x):
+                                        im_sol += Interval.Lopen(start, z)
+                                    else:
+                                        im_sol += Interval.open(start, z)
+                            start = z
```
`valid_start = valid_z`

akash9712 committed:
solve_univariate_inequality: Changes in checks on critical points

Validity of critical points is computed once now, and stored in  a variable
instead of calling the `valid` function every time.

akash9712 commented:
@smichr Made the changes as recommened.

akash9712 committed:
Changes to solveset for `arg`

Changed solveset to avoide returning wrong values if `arg` is being equated
to real values other than zero.

smichr committed:
refactor use of valid_* variables

akash9712 commented:
Thanks!
